### PR TITLE
feat: dynamic resizing of grid items

### LINKED ISSUE #1036: Implement dynamic resizing for profile layout sections
### What would you like to be added?

Dynamically measure the size of profile sections and update the `layouts` of `react-grid-layouts`
Improves upon #1012 

### Why is this needed?

`react-grid-layouts` by default only supports fixed widths and heights for sections, and consequently default profiles of many players has a lot of empty spaces. 

How it is atm:
Screenshot 2022-01-05 at 16 22 24

How it should be:
Screenshot 2022-01-05 at 16 24 27

### PR DESCRIPTION
**Pull request by dan13ram**
## Overview

**What features/fixes does this PR include?**
This should fix the issue where the fixed height in `react-grid-layout` was raising problems when the content of the box was either too short or too long.

**Please provide the GitHub issue number**

Closes #1036 

## Follow up Improvement Ideas

- [ ] please list any improvement/ideas

## Implementation

**Describe technical (nontrivial / non-obvious) parts of your code**
It uses a ResizeObserver to observer for height changes and update the layout accordingly.

**Side effects**

## Assets

[Include screenshots/videos if it makes reviewing easier.]

### CODE REVIEW

**Review by vidvidvid — COMMENTED** [review]
Lovely! I tested editing the layout and everything seems to be working fine! :)
I just found one tiny nuisance. When adding an empty block, the size of it is too large, but it gets set to the right size after refreshing the page:
block
 
Could this be solved as a part of this PR?

**Comment by dan13ram** [comment]
Thank you @vidvidvid I'll take a look at that issue. I'll solve it within the PR

**Comment by vidvidvid** [comment]
One more thing - there is a gap between the cover photo and the megamenu, idk how I didn't notice this earlier 😅 
Screenshot 2022-01-21 at 21 38 56


And while you're at it - if not reverting back to how the profile hero was positioned before the layout upgrade - might we consider removing or at least decreasing the padding on top? I think that the cover photo takes too much space now. @davort could we have your thoughts on this?
padding_profile

**Comment by dan13ram** [comment]
@vidvidvid I've fixed that issue you mentioned and also removed the gap on top. I've kept the padding on top so that the cover photo remains visible.

**Comment by vidvidvid** [comment]
But maybe we should wait until @dysbulic merges his pr so he won't have to do the rebasing yet again. @dysbulic what do you say?

**Comment by vidvidvid** [comment]
Although it doesn't have any conflicts with `develop`, so there might be no conflicts when rebasing.

**Comment by dan13ram** [comment]
I understand @vidvidvid We can wait for @dysbulic

**Comment by dysbulic** [comment]
> But maybe we should wait until @dysbulic merges his pr so he won't have to do the rebasing yet again.

I'd sure appreciate it. I definitely want to see this one go through, but hopefully we can get a review done in the [meeting tomorrow](//discord.gg/Sy7jupey?event=933387177990168656) and get #943 merged.

File: packages/web/pages/player/[username].tsx [hunk 1]
```diff
@@ -178,37 +245,51 @@ export const Grid: React.FC<Props> = ({ player: initPlayer }): ReactElement => {
...
+          },
+    [player?.profile_layout],
   );
 
   const [
     { layoutItems: currentLayoutItems, layouts: currentLayouts },
     setCurrentLayoutData,
   ] = useState<ProfileLayoutData>(savedLayoutData);
 
+  const itemsRef = useRef<(HTMLDivElement | null)[]>([]);
```
Comment (dysbulic) [inline packages/web/pages/player/[username].tsx:264]:
```suggestion
  const itemsRef = useRef<Array<HTMLElement>>([]);
```

File: packages/web/pages/player/[username].tsx [hunk 2]
```diff
@@ -178,37 +245,51 @@ export const Grid: React.FC<Props> = ({ player: initPlayer }): ReactElement => {
...
+
+  useEffect(() => {
+    itemsRef.current = itemsRef.current.slice(0, currentLayoutItems.length);
+  }, [currentLayoutItems]);
+
+  const heights = useItemHeights(itemsRef, currentLayoutItems);
```
Comment (dysbulic) [inline packages/web/pages/player/[username].tsx:270]:
```suggestion
  const heights = useItemHeights(itemsRef.current, currentLayoutItems);
```

File: packages/web/pages/player/[username].tsx [hunk 3]
```diff
@@ -427,50 +499,47 @@ export const Grid: React.FC<Props> = ({ player: initPlayer }): ReactElement => {
...
           md: [20, 20],
           sm: [20, 20],
-          xxs: [15, 15],
         }}
       >
-        {currentLayoutItems.map(({ boxKey, boxType, boxMetadata }) =>
+        {currentLayoutItems.map(({ boxKey, boxType, boxMetadata }, i) =>
           boxType === BoxType.PLAYER_ADD_BOX ? (
-            <Flex key={boxKey} className="gridItem">
+            <Flex key={boxKey} className="gridItem" id={boxKey}>
```
Comment (dysbulic) [inline packages/web/pages/player/[username].tsx:521]: Why are you setting the `id`? Is it referenced by a stylesheet or something?

Also, the `Flex` tags for these two sections are the same, they can be moved to wrap the ternary and deduplicated.

**Review by dysbulic — COMMENTED** [review]
Lots of little formatting quibbles, but the basic layout seems to make sense.

File: packages/web/pages/player/[username].tsx (hunk 3 above)
Comment (dan13ram) [reply packages/web/pages/player/[username].tsx]: @dysbulic The `id` is needed for `getBoxKeyFromTarget` in line 166

**Comment by dan13ram** [comment]
@dysbulic Do take another look now if this is good to merge :)

**Pull request merged**